kubernetes/resource_kubernetes_env.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	gvk, name, namespace, err := util.ParseResourceID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// figure out which resource client to use
+	dc, err := m.(KubeClientsets).DiscoveryClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	agr, err := restmapper.GetAPIGroupResources(dc)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	restMapper := restmapper.NewDiscoveryRESTMapper(agr)
+	mapping, err := restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// determine if the resource is namespaced or not
+	var r dynamic.ResourceInterface
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		if namespace == "" {
+			namespace = "default"
+		}
+		r = conn.Resource(mapping.Resource).Namespace(namespace)
+	} else {
+		r = conn.Resource(mapping.Resource)
+	}
+
+	// get the resource environments
+	res, err := r.Get(ctx, name, v1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return diag.Diagnostics{{
+				Severity: diag.Warning,
+				Summary:  "Resource deleted",
+				Detail:   fmt.Sprintf("The underlying resource %q has been deleted. You should recreate the underlying resource, or remove it from your configuration.", name),
+			}}
+		}
+		return diag.FromErr(err)
+	}
+
+	// store names of environment variables into map
+	configuredEnvs := make(map[string]interface{})
+	envList := d.Get("env").([]interface{})
+	for _, e := range envList {
+		configuredEnvs[e.(map[string]interface{})["name"].(string)] = ""
+	}
+
+	// strip out envs not managed by Terraform
+	managedEnvs, err := getManagedEnvs(res.GetManagedFields(), defaultFieldManagerName, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	responseEnvs, err := getResponseEnvs(res, d.Get("container").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	env := []interface{}{}
+	for _, e := range responseEnvs {
+		envName := e.(map[string]interface{})["name"].(string)
+		_, managed := managedEnvs[fmt.Sprintf(`k:{"name":%q}`, envName)]
+		_, configured := configuredEnvs[envName]
+		if !managed && !configured {
+			continue
+		}
+		env = append(env, e)
+	}
+
+	// envs := []interface{}{
+	// 	map[string]interface{}{
+	// 		"name": "NGINX_HOST",
+	// 		"value": "foobar.com"},
+	// 	map[string]interface{}{
+	// 		"name": "NGINX_PORT",
+	// 		"value": "90"},
+	// 	map[string]interface{}{
+	// 		"name": "EXPORTED_VARIABLE_FROM_SECRET",
+	// 		"valueFrom": map[string]interface{}{
+	// 			"secretKeyRef": map[string]interface{}{
+	// 				"key": "one",
+	// 				"name": "tf-acc-test-4092216009068675976",
+	// 				"optional": true
+	// 			}}},
+	// 	map[string]interface{}{
+	// 		"name": "EXPORTED_VARIABLE_FROM_CONFIG_MAP",
+	// 		"valueFrom": map[string]interface{}{
+	// 			"configMapKeyRef": map[string]interface{}{
+	// 				"key": "one",
+	// 				"name": "tf-acc-test-271185184630974840",
+	// 				"optional": true}}}}
+
+	env, _ = flattenEnv(env)
```

Comment:
the `flattenEnv` doesn't actually return errors anywhere, so change the function just to return a map only.